File: lab1/fp_SimUDuck/Duck.cpp
#include "stdafx.h"
#include "Duck.h"

Duck::Duck(std::function<void()>&& flyBehavior,
	std::function<void()>&& quackBehavior,
	std::function<void()>&& danceBehavior)
	:m_quackBehavior(std::move(quackBehavior))
	,m_danceBehavior(std::move(danceBehavior))
{
	assert(m_quackBehavior);
	assert(m_danceBehavior);
	SetFlyBehavior(std::move(flyBehavior));
}

void Duck::Quack() const
{
	m_quackBehavior();
}

void Duck::Swim()
{
	std::cout << "I'm swimming" << std::endl;
}

void Duck::Fly()
{
	m_flyBehavior();
}

void Duck::Dance()
{
	m_danceBehavior();
}

void Duck::SetFlyBehavior(std::function<void()>&& flyBehavior)
{
	assert(flyBehavior);
	m_flyBehavior = std::move(flyBehavior);
}

File: lw1_fn/src/duck/Duck.php
<?php

abstract class Duck
{
    /** @var Closure */
    private $flyBehavior;
    /** @var Closure */
    private $quackBehavior;
    /** @var Closure */
    private $danceBehavior;

    public function __construct(callable $flyBehavior,
                                callable $quackBehavior,
                                callable $danceBehavior)
    {
        $this->flyBehavior = $flyBehavior;
        $this->quackBehavior = $quackBehavior;
        $this->danceBehavior = $danceBehavior;
    }

    public function quack(): void
    {
        call_user_func($this->quackBehavior);
    }

    public function swim(): void
    {
        echo 'I\'m swimming' . PHP_EOL;
    }

    public function fly(): void
    {
        call_user_func($this->flyBehavior);
    }

    public function dance(): void
    {
        call_user_func($this->danceBehavior);
    }

    public function setFlyBehavior(callable $flyBehavior)
    {
        $this->flyBehavior = $flyBehavior;
    }

    abstract public function display(): void;
}